### DISCUSSION TITLE
Lazer renders next notes behind previous, differing to stable.

### DISCUSSION THREAD

**Discussion by yuuiko** [ppy/osu#21994] [Ideas]
## Edit: please look at reply below. This issue is across modes
https://github.com/ppy/osu/discussions/21994#discussioncomment-4874452

----

The below used to be a suggestion to enhance mania readability (stable renders next notes behind previous on mania too, in most cases. I was suggesting for an option to be reversed in lazer). The issue has since changed.

It can be hard to see dense stacked notes.
Suggest to render all next notes on top of the previous note, instead of underneath as it is currently.

<img width="426" alt="Screenshot 2023-01-02 at 12 25 32 pm" src="https://user-images.githubusercontent.com/50823728/210189750-62451c08-37eb-4f6b-adb2-1884af6612e1.png">
in the image above, you can see the slider end being hidden behind the slider head.

**Comment by frenzibyte** [comment]
Please link the beatmap, it's hard to read what's going on in the screenshot you attached.

**Reply by yuuiko** [reply]
well thats the problem, it's hard to read whats going on in dense mania maps

https://osu.ppy.sh/beatmapsets/1483944#mania/3043450 is the map but happens on any map really

**Reply by frenzibyte** [reply]
I mean, we need to know the map and timestamp of the screenshot above, to understand why argon is behaving badly and improve it. As I see it, it looks as if you're playing at a lower scroll speed here and that causes the hold note and notes to conflict each other.

**Reply by bdach** [reply]
> it looks as if you're playing at a lower scroll speed here

I mean not necessarily. The point - if I understand correctly - is that with two argon notes close together, you can't tell if it's a short hold or two notes. It's probably exaggerated on low scroll maps for sure but there'll inevitably be gimmick maps where this looks bad too regardless of scroll speed.

**Reply by yuuiko** [reply]
> I mean, we need to know the map and timestamp of the screenshot above, to understand why argon is behaving badly and improve it. As I see it, it looks as if you're playing at a lower scroll speed here and that causes the hold note and notes to conflict each other.

Happens with any map at any scroll speed but yes it is most visible on sliders or low scroll speeds.

Seperate issue to https://github.com/ppy/osu/discussions/21996

**Reply by peppy** [reply]
To confirm, you're citing an issue *only* with the argon skin right?

**Reply by yuuiko** [reply]
Not really, could be a problem with different skins.
Argon just has extra unreadability due to its design, check @Bruno5430 reply below

**Comment by Bruno5430** [comment]
I've been comparing how it behaves compared to stable and while in stable hit notes are randomly layered, i don't think it will improve readability if the next notes render on top of the previous ones, lazer already solves the random layering by rendering the next notes underneath the previous ones
This is not the case for hold bodies or tails, but i've opened an issue for that in #22272 because it's more clear what the layering should be.
There's also visual mods like flashlight which needs to be considered because it's common for players to reduce scroll speed while playing with flashlight.
I played around flashlight a little bit, but i would prefer to hear feedback from other flashlight players about the layering
| Lazer | Stable |
| --- | --- |
| (685ms scroll speed) | (17 scroll speed) |
| ![osu_2023-01-18_00-26-53](https://user-images.githubusercontent.com/111712756/213076450-306bb2b3-85f8-4bcb-86d4-739dac06c4cd.png) | ![screenshot143](https://user-images.githubusercontent.com/111712756/213075200-35436971-cd06-4f04-b6fe-0c321ef58833.jpg) |
[Beatmap: Glaive - Astrid](https://osu.ppy.sh/beatmapsets/1805498#mania/3702593)

**Reply by Walavouchey** [reply]
for clarity, i want to point out that while the note layering may be random in stable, hold note bodies are always drawn below regular notes (related to my top-level comment below)

copying the comparison in https://github.com/ppy/osu/issues/22272:

![image](https://user-images.githubusercontent.com/36758269/216823261-daa3ca02-7059-43c6-b25e-292ce782cc91.png)

**Comment by Walavouchey** [comment]
this isn't a mania-specific issue btw

below is a 2b map to illustrate that in stable (osu! game mode), slider bodies are always drawn below slider heads, slider tails and hit circles, basically on a separate layer.

lazer with argon, classic, my skin, and triangles:

https://user-images.githubusercontent.com/36758269/216821377-8d61f57b-df14-4ac5-b328-64638272b91d.mp4

stable with default skin:

https://user-images.githubusercontent.com/36758269/216821919-f771bd85-11f6-485b-92e4-c7255ee65484.mp4

**Reply by bdach** [reply]
this is a separate discussion because both issues can't be both fixed at once with a single code change.

i will also say that stable's z-ordering code is very wacky to the point of possibly being irreproducible given lazer's structure. see https://github.com/ppy/osu/pull/14561#issuecomment-907785826 for prior notes on this

**Reply by Walavouchey** [reply]
moving the material i've posted here to https://github.com/ppy/osu/discussions/14594, then.

**Comment by yuuiko** [comment]
## Here is a mockup of what I mean

This mockup is part of improving mania readability: [#21996 ](https://github.com/ppy/osu/discussions/21996#discussioncomment-4929241)

Swap the render mode so that each next note is rendered in front of the previous instead of behind
This will break readability on some skins that were made for the "next is behind" in mind though (made with trailing shadows instead of leading shadows), so make it a skinning option.

<img width="689" alt="Screenshot 2023-02-16 at 8 56 14 pm" src="https://user-images.githubusercontent.com/50823728/219332564-71a7a194-103c-474b-9955-8dfa50349c1e.png">
